File: app/src/main/java/com/petnbu/petnbu/feed/FeedsViewModel.java
package com.petnbu.petnbu.feed;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.Nullable;

import com.petnbu.petnbu.PetApplication;
import com.petnbu.petnbu.SingleLiveEvent;
import com.petnbu.petnbu.model.FeedUI;
import com.petnbu.petnbu.model.Paging;
import com.petnbu.petnbu.model.Feed;
import com.petnbu.petnbu.model.Resource;
import com.petnbu.petnbu.repo.FeedRepository;
import com.petnbu.petnbu.repo.LoadMoreState;
import com.petnbu.petnbu.repo.UserRepository;

import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

public class FeedsViewModel extends ViewModel {

    @Inject
    FeedRepository mFeedRepository;

    @Inject
    UserRepository mUserRepository;

    @Inject
    Application mApplication;

    private final SingleLiveEvent<String> mOpenUserProfileEvent = new SingleLiveEvent<>();
    private final SingleLiveEvent<String> mOpenCommentsEvent = new SingleLiveEvent<>();

    private LiveData<Resource<List<FeedUI>>> mFeedsLiveData;

    private LoadMoreHandler loadMoreHandler;

    public FeedsViewModel() {
        PetApplication.getAppComponent().inject(this);
        loadMoreHandler = new LoadMoreHandler(mFeedRepository);
    }

    public LiveData<Resource<List<FeedUI>>> getFeeds(String pagingId, String loggedUserId) {
        mFeedsLiveData = mFeedRepository.loadFeeds(pagingId, loggedUserId);
        return mFeedsLiveData;
    }

    public LiveData<LoadMoreState> getLoadMoreState() {
        return loadMoreHandler.getLoadMoreState();
    }

    public void loadNextPage() {
        if (mFeedsLiveData.getValue() != null) {
            loadMoreHandler.loadNextPage(Paging.GLOBAL_FEEDS_PAGING_ID);
        }
    }

    @Override
    protected void onCleared() {
        super.onCleared();
    }

    public LiveData<Resource<List<Feed>>> refresh() {
        loadMoreHandler.reset();
        return mFeedRepository.refresh();
    }

    public void openUserProfile(String userId) {
        mOpenUserProfileEvent.setValue(userId);
    }

    public LiveData<String> getOpenUserProfileEvent() {
        return mOpenUserProfileEvent;
    }

    public void openComments(String feedId) {
        mOpenCommentsEvent.setValue(feedId);
    }

    public LiveData<String> getOpenCommentsEvent() {
        return mOpenCommentsEvent;
    }

    static class LoadMoreHandler implements Observer<Resource<Boolean>> {

        private final FeedRepository feedRepo;

        private MutableLiveData<LoadMoreState> loadMoreState = new MutableLiveData<>();

        private LiveData<Resource<Boolean>> nextPageLiveData;

        private boolean hasMore = true;

        public LoadMoreHandler(FeedRepository feedRepository) {
            feedRepo = feedRepository;
            loadMoreState.setValue(new LoadMoreState(false, null));
        }

        public void loadNextPage(String pagingId) {
            if (!hasMore || loadMoreState.getValue() == null || loadMoreState.getValue().isRunning()) {
                Timber.i("hasMore = %s", hasMore);
                return;
            }
            Timber.i("loadNextPage");
            unregister();
            nextPageLiveData = feedRepo.fetchNextPage(pagingId);
            loadMoreState.setValue(new LoadMoreState(true, null));
            nextPageLiveData.observeForever(this);

        }

        @Override
        public void onChanged(@Nullable Resource<Boolean> result) {
            if (result == null) {
                reset();
            } else {
                Timber.i(result.toString());
                switch (result.status) {
                    case SUCCESS:
                        hasMore = Boolean.TRUE.equals(result.data);
                        unregister();
                        loadMoreState.setValue(new LoadMoreState(false, null));
                        break;
                    case ERROR:
                        hasMore = true;
                        unregister();
                        loadMoreState.setValue(new LoadMoreState(false,
                                result.message));
                        break;
                }
            }
        }

        private void unregister() {
            if (nextPageLiveData != null) {
                nextPageLiveData.removeObserver(this);
                nextPageLiveData = null;
            }
        }

        private void reset() {
            unregister();
            hasMore = true;
            loadMoreState.setValue(new LoadMoreState(false, null));
        }

        public MutableLiveData<LoadMoreState> getLoadMoreState() {
            return loadMoreState;
        }
    }
}

File: app/src/main/java/com/petnbu/petnbu/feed/FeedsViewModel.kt
package com.petnbu.petnbu.feed

import android.arch.lifecycle.*
import com.petnbu.petnbu.PetApplication
import com.petnbu.petnbu.SingleLiveEvent
import com.petnbu.petnbu.model.*
import com.petnbu.petnbu.repo.FeedRepository
import com.petnbu.petnbu.repo.LoadMoreState
import timber.log.Timber
import javax.inject.Inject

class FeedsViewModel : ViewModel() {

    @Inject
    lateinit var feedRepository: FeedRepository

    val showLoadingErrorEvent = SingleLiveEvent<String>()
    val showLoadingEvent = SingleLiveEvent<Boolean>()
    val openUserProfileEvent = SingleLiveEvent<String>()
    val openCommentsEvent = SingleLiveEvent<String>()

    private val feedsLiveData: MutableLiveData<List<FeedUI>> = MutableLiveData()
    private val loadMoreHandler: LoadMoreHandler

    val loadMoreState: LiveData<LoadMoreState>
        get() = loadMoreHandler.loadMoreState

    init {
        PetApplication.appComponent.inject(this)
        loadMoreHandler = LoadMoreHandler(feedRepository)
    }

    fun getFeeds(pagingId: String, loggedUserId: String): LiveData<List<FeedUI>> {
        return Transformations.switchMap(feedRepository.loadFeeds(pagingId, loggedUserId), {
            it?.run {
                data?.run {
                    feedsLiveData.value = this
                }
                showLoadingEvent.value = status == Status.LOADING
                takeIf { status == Status.ERROR }?.run {
                    showLoadingErrorEvent.value = message
                }
            }
            feedsLiveData
        })
    }

    fun loadNextPage() {
        feedsLiveData.value?.run {
            loadMoreHandler.loadNextPage(Paging.GLOBAL_FEEDS_PAGING_ID)
        }
    }

    fun refresh() {
        loadMoreHandler.reset()
        val refreshLiveData = feedRepository.refresh()
        refreshLiveData.observeForever(object :Observer<Resource<List<Feed>>> {
            override fun onChanged(resource: Resource<List<Feed>>?) {
                if(resource != null && resource.status != Status.LOADING) {
                    showLoadingEvent.value = false
                    refreshLiveData.removeObserver(this)
                }
            }
        })
    }

    fun openUserProfile(userId: String) {
        openUserProfileEvent.value = userId
    }

    fun openComments(feedId: String) {
        openCommentsEvent.value = feedId
    }

    fun likeClicked(userId: String, feedId: String) {
        feedRepository.likeFeedHandler(userId, feedId)
    }

    private class LoadMoreHandler(private val feedRepo: FeedRepository,
                                  val loadMoreState: MutableLiveData<LoadMoreState> = MutableLiveData())
        : Observer<Resource<Boolean>> {

        private lateinit var nextPageLiveData: LiveData<Resource<Boolean>>

        private var hasMore = true

        init {
            loadMoreState.value = LoadMoreState(false, null)
        }

        fun loadNextPage(pagingId: String) {
            if (!hasMore || loadMoreState.value == null || loadMoreState.value!!.isRunning) {
                Timber.i("hasMore = %s", hasMore)
                return
            }
            Timber.i("loadNextPage")
            unregister()
            nextPageLiveData = feedRepo.fetchNextPage(pagingId)
            loadMoreState.value = LoadMoreState(true, null)
            nextPageLiveData.observeForever(this)
        }

        override fun onChanged(result: Resource<Boolean>?) {
            if (result == null) {
                reset()
            } else {
                Timber.i(result.toString())
                when (result.status) {
                    Status.SUCCESS -> {
                        hasMore = java.lang.Boolean.TRUE == result.data
                        unregister()
                        loadMoreState.setValue(LoadMoreState(false, null))
                    }
                    Status.ERROR -> {
                        hasMore = true
                        unregister()
                        loadMoreState.setValue(LoadMoreState(false,
                                result.message ?:"Error Unknown"))
                    }
                }
            }
        }

        private fun unregister() {
            if (this::nextPageLiveData.isInitialized) {
                nextPageLiveData.removeObserver(this)
            }
        }

        fun reset() {
            unregister()
            hasMore = true
            loadMoreState.value = LoadMoreState(false, null)
        }
    }
}
